File: coding/solutions/src/main/java/xyz/grind/coding/blind75/TopKFrequentElements.java
package xyz.grind.coding.blind75;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;

public class TopKFrequentElements {

   public static int[] topKFrequentElementsBruteForce(int[] nums, int k) {
      if(nums.length == k) return nums;

      Map<Integer, Integer> map = new HashMap<>();
      for (int i : nums) map.put(i, map.getOrDefault(i, 0) + 1);

      List<Map.Entry<Integer, Integer>> list = new ArrayList<>(map.entrySet());

       list.sort(Map.Entry.comparingByValue((a, b) -> b - a));

       int[] top = new int[k];
       for (int i=0; i < k; i++) {
          top[i] = list.get(i).getKey();
       }

      return top;
   }

   public static int[] topKFrequentElementsHeap(int[] nums, int k){
      // O(1) time
      if (k == nums.length) {
         return nums;
      }

      // 1. build hash map : character and how often it appears
      // O(N) time
      Map<Integer, Integer> count = new HashMap();
      for (int n: nums) {
         count.put(n, count.getOrDefault(n, 0) + 1);
      }

      // init heap 'the less frequent element first'
      Queue<Integer> heap = new PriorityQueue<>(
              (n1, n2) -> count.get(n1) - count.get(n2));

      // 2. keep k top frequent elements in the heap
      // O(N log k) < O(N log N) time
      for (int n: count.keySet()) {
         System.out.println("keyset " + n);
         heap.add(n);
         if (heap.size() > k) heap.poll();
      }

      // 3. build an output array
      // O(k log k) time
      int[] top = new int[k];
      for(int i = k - 1; i >= 0; --i) {
         top[i] = heap.poll();
      }
      return top;
   }



   public static int[] topKFrequentElements(int[] nums, int k) {
      if(k == nums.length) return nums;

      Map<Integer, Integer> counts = new HashMap<Integer, Integer>();
      for(int num : nums) {
         counts.put(num, counts.getOrDefault(num, 0) + 1);
      }


      List<Integer>[] buckets = new List[nums.length + 1];
      for(int i = 0; i < buckets.length; i++) {
         buckets[i] = new ArrayList<>();
      }


      for(int count: counts.keySet()) {
         buckets[counts.get(count)].add(count);
      }

      List<Integer> flattened = new ArrayList<>();
      for(int i = buckets.length - 1; i >= 0; i--) {
         for(int element : buckets[i]) {
            flattened.add(element);
         }
      }

      int[] top = new int[k];
      for(int i = 0; i < k; i++) {
         top[i] = flattened.get(i);
      }

      return top;
   }


}

File: problem-solving/solutions/leetcode/blind75/src/main/kotlin/xyz/grind/coding/easy/TopKFrequentElements.kt
package xyz.grind.coding.easy

import java.util.PriorityQueue
import java.util.Queue
import kotlin.collections.MutableList
import kotlin.collections.set

object TopKFrequentElements {

    fun topKFrequentElementsHeap(nums: IntArray, k: Int): IntArray {
        if (k == nums.size) {
            return nums
        }
        val count = mutableMapOf<Int, Int>()
        for (n in nums) { count[n] = count.getOrDefault(n, 0) + 1 }

        // when initializing the heap we keep the less frequent element first
        val heap: Queue<Int> = PriorityQueue { n1, n2 -> count[n1]!! - count[n2]!! }
        for (n in count.keys) {
            heap.add(n)
            if (heap.size > k) heap.poll()
        }
        val top = IntArray(k)
        for (i in k - 1 downTo 0) top[i] = heap.poll()
        return top
    }

    fun topKFrequentElements(nums: IntArray, k: Int): IntArray {
        val count = hashMapOf<Int, Int>()
        val freq = MutableList<MutableList<Int>>(nums.size + 1) { mutableListOf() }
        for (n in nums) count[n] = count.getOrDefault(n, 0) + 1
        for ((n, c) in count) { freq[c].add(n) }

        val res = mutableListOf<Int>()
        for (i in freq.size - 1 downTo 0) {
            for (n in freq[i]) {
                res.add(n)
                if (res.size == k) return res.toIntArray()
            }
        }
        return intArrayOf()
    }
}
